addons/website_sale/controllers/main.py

```diff
@@ -272,8 +272,8 @@ def product(self, product, category='', search='', **kwargs):
 
         if not context.get('pricelist'):
             context['pricelist'] = int(self.get_pricelist())
-            product = template_obj.browse(cr, uid, int(product), context=context)
-
+            # Public user should able to read virtual stock
+            product = template_obj.browse(cr, SUPERUSER_ID, int(product), context=context)
```

Comment:
I think to make the function which compute the `qty_available`of product as sudo, but I'm afraid it break lots of things, so for now, add access right for public for `stock.move, stock.warehouse, stock.quants, stock.location, product.uom`, and remove the SUPERUSER_ID above.